opentelemetry-user-events-metrics/src/exporter/mod.rs

```diff
@@ -46,411 +89,514 @@ impl Debug for MetricsExporter {
+        // ═══════════════════════════════════════════════════════════════════════════════════
+        // Iterate through each data point within this exponential histogram metric.
+        // Each data point = unique combination of (metric + attributes + timestamp + buckets).
+        // SERIALIZATION: Each data point is individually encoded and emitted to tracepoint.
+        for dp in hist.data_points() {
+            let histogram_data_point = opentelemetry_proto::tonic::metrics::v1::ExponentialHistogramDataPoint {
+                attributes: dp.attributes().map(Into::into).collect(),
+                start_time_unix_nano: hist_start_time,
+                time_unix_nano: hist_time,
+                count: dp.count() as u64,
+                sum: Some(dp.sum().into_f64()),
+                scale: dp.scale().into(),
+                zero_count: dp.zero_count(),
+                positive: Some(opentelemetry_proto::tonic::metrics::v1::exponential_histogram_data_point::Buckets {
+                    offset: dp.positive_bucket().offset(),
+                    bucket_counts: dp.positive_bucket().counts().collect(),
+                }),
+                negative: Some(opentelemetry_proto::tonic::metrics::v1::exponential_histogram_data_point::Buckets {
+                    offset: dp.negative_bucket().offset(),
+                    bucket_counts: dp.negative_bucket().counts().collect(),
+                }),
+                exemplars: Vec::new(), // No support for exemplars
+                flags: default_flags,
+                min: dp.min().map(|v| v.into_f64()),
+                max: dp.max().map(|v| v.into_f64()),
+                zero_threshold: dp.zero_threshold(),
+            };
+
+            let histogram_point_proto =
+                opentelemetry_proto::tonic::metrics::v1::ExponentialHistogram {
+                    aggregation_temporality: 1,
+                    data_points: vec![histogram_data_point],
+                };
+
+            // Update the data field for this data point
+            export_metric_service_request_common.resource_metrics[0].scope_metrics[0].metrics[0]
+                .data = Some(
+                opentelemetry_proto::tonic::metrics::v1::metric::Data::ExponentialHistogram(
+                    histogram_point_proto,
+                ),
+            );
+
+            byte_array.clear(); // Clear contents but retain capacity for performance
+            if self
+                .encode_and_emit_metric(export_metric_service_request_common, byte_array, metric)
+                .is_err()
+            {
+                failed_count += 1;
+            }
+        }
+        failed_count
+    }
+
+    fn encode_and_emit_metric(
+        &self,
+        export_metric_service_request_common: &ExportMetricsServiceRequest,
+        byte_array: &mut Vec<u8>,
+        metric: &opentelemetry_sdk::metrics::data::Metric,
+    ) -> Result<(), String> {
+        match export_metric_service_request_common.encode(byte_array) {
```

Comment:
You don't have to actually encode the message to find out if the length exceeds 64KB. Use `Message:encoded_len()` to calculate the size of the serialized bytes upfront. Refer to ETW exporter for this:

https://github.com/open-telemetry/opentelemetry-rust-contrib/blob/9741f10aaafe7e96a51e33cab0fa2030cb73d603/opentelemetry-etw-metrics/src/exporter/mod.rs#L48-L50

Reply:
nice. Don't see this making a big difference in current implementation, as possibility of hitting the limit with just one datapoint is very low. But this will come in handy when we try to put multiple datapoints into one tracepoint::write!

Reply:
Yeah, it might also be useful with exporting exemplars.